### PR TITLE
[9/n] [dropshot] add API traits with support for endpoints

### PR DESCRIPTION
**Pull request by sunshowers**
As discussed in [RFD 479](https://rfd.shared.oxide.computer/rfd/0479) and https://github.com/oxidecomputer/omicron/issues/5247.

Add support for a new way to define Dropshot APIs: via a trait. This builds on work done in the prior PRs in this series, such that it reuses as much shared logic from endpoints as possible.

I've tried to structure the PR series such that this one is almost entirely added code. It's also mostly tests, with around 1400 lines of "real" code.

See RFD 479 for more information.

### TODO

- [x] Add support for channels: #1038
- [x] Handle all XXXs
- [x] Cover failure test cases to mirror endpoint ones
- [x] Add more success tests
- [ ] Answer open questions in RFD

### CODE REVIEW

File: dropshot/src/lib.rs [hunk 1]
```diff
@@ -682,4 +683,9 @@ pub use http::Method;
 
 extern crate dropshot_endpoint;
 pub use dropshot_endpoint::channel;
+pub use dropshot_endpoint::dropshot_server;
 pub use dropshot_endpoint::endpoint;
+
+// Exported for dropshot_endpoint -- not part of the public API.
+#[doc(hidden)]
+pub use trait_variant::make as make_trait_variant;
```
Comment (sunshowers) [inline dropshot/src/lib.rs:691]: Using `trait-variant` temporarily, but it does too much -- we only care about making our endpoints `Send`able. We should copy whatever code from trait-variant we need.

Comment (sunshowers) [reply dropshot/src/lib.rs]: No longer use trait-variant, inlining the bits we needed was just around 15 lines of code.

File: dropshot/src/lib.rs [hunk 2]
```diff
@@ -178,6 +178,83 @@
 //! for each type of response (which can also include documentation).  This is
 //! largely known statically, though generated at runtime.
 //!
+//! ### As a trait server
```
Comment (sunshowers) [inline dropshot/src/lib.rs:181]: This is where the end-user documentation starts.

File: dropshot/tests/fail/bad_server_invalid_types.stderr [hunk 3]
```diff
@@ -0,0 +1,209 @@
+error[E0432]: unresolved import `dropshot::dropshot_server`
```
Comment (sunshowers) [inline dropshot/tests/fail/bad_server_invalid_types.stderr:1]: This file needs to be deleted.

File: dropshot/tests/fail/bad_server_endpoint26.stderr [hunk 4]
```diff
@@ -0,0 +1,7 @@
+error: endpoint `bad_endpoint` has invalid attributes: missing field `path`
+  --> tests/fail/bad_server_endpoint26.rs:12:1
+   |
+12 | #[dropshot::server]
+   | ^^^^^^^^^^^^^^^^^^^
+   |
+   = note: this error originates in the attribute macro `dropshot::server` (in Nightly builds, run with -Z macro-backtrace for more info)
```
Comment (sunshowers) [inline dropshot/tests/fail/bad_server_endpoint26.stderr:7]: This is addressed by https://github.com/oxidecomputer/serde_tokenstream/pull/194.

**Review by ahl — APPROVED** [review]
I'd say I looked that code "some" i.e. in the places that I felt most familiar with. I scanned through the tests to think about coverage, and I looked at the docs for end-user clarity. If you'd like me to look more thoroughly at some part(s) please let me know but my general feeling is that the mechanics don't require intense scrutiny.

File: dropshot/examples/server-trait-default.rs
[the remark this answers is not in the document]
Comment (sunshowers) [reply dropshot/examples/server-trait-default.rs:1]: How about `server-trait-alternate.rs`? Definitely hear you about `default` being misleading in this context.

> Also: is there a particular situation in which one would build a server this way?

I'm not entirely sure how the patterns will play out to be honest! I think there's definitely at least some reason to build a server this way, providing an HTTP layer over an application interface. I'm curious to see how things shake out in practice.

File: dropshot/examples/server-trait-default.rs [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+//! An extended version of `server-trait.rs`, demonstrating use of the
+//! `dropshot::server` attribute macro to define a server with default methods.
+//!
+//! In this example, all the behavior lives on the context type, and the
+//! Dropshot server is defined as a trait with default methods that delegate to
+//! the context type.
+//!
+//! There are a number of possible variations of this, including:
+//!
+//! * The base behavior and endpoints are methods on the same trait, allo=wing
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply dropshot/examples/server-trait-default.rs:12]: Fixed, thanks.

File: dropshot/examples/server-trait-default.rs [hunk 6]
```diff
@@ -0,0 +1,166 @@
+// Copyright 2024 Oxide Computer Company
+
+//! An extended version of `server-trait.rs`, demonstrating use of the
+//! `dropshot::server` attribute macro to define a server with default methods.
+//!
+//! In this example, all the behavior lives on the context type, and the
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply dropshot/examples/server-trait-default.rs:6]: "server logic" is good, will change to that.

File: dropshot/examples/server-trait-default.rs [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+//! the context type.
+//!
+//! There are a number of possible variations of this, including:
+//!
+//! * The base behavior and endpoints are methods on the same trait, allo=wing
+//!   implementations to override the base behavior if desired (similar to std's
+//!   `Read` and `Write`).
+//! * The behavior lives in a base trait, and the server is an extension trait
+//!   on top of it, with a blanket impl prohibiting overrides (similar to
+//!   tokio's `AsyncReadExt` and `AsyncWriteExt`).
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply dropshot/examples/server-trait-default.rs:17]: Expanded the doc comment to explain this in more detail.

## Variations

The general degrees of freedom available when dealing with traits in Rust
are almost all available here. The main bit of flexibility that isn't
available is that server traits can't be object-safe.

### Same trait with default methods

For example:

```rust,ignore
#[dropshot::server]
trait CounterServer {
     type Context;
     
     fn get_counter_impl(&self) -> impl Future<Output = u64> + Send;
     fn set_counter_impl(
         &self,
         value: u64,
     ) -> impl Future<Output = Result<(), String>> + Send;
     #[endpoint { method = GET, path = "/counter" }
     async fn get_counter() {} // as below
     #[endpoint { method = PUT, path = "/counter" }
     async fn put_counter() {} // as below
}
```

This is similar to how [`std::io::Write`]'s `write_all` method is defined as
a default method over the required `write`.

### Supertrait

```rust,ignore
trait CounterBase {
    fn get_counter_impl(&self) -> impl Future<Output = u64> + Send;
    fn set_counter_impl(
        &self,
        value: u64,
    ) -> impl Future<Output = Result<(), String>> + Send;
}
#[dropshot::server]
trait CounterServer: CounterBase {
    #[endpoint { method = GET, path = "/counter" }
    async fn get_counter() {} // as below
    #[endpoint { method = PUT, path = "/counter" }
    async fn put_counter() {} // as below
}
// And optionally, a blanket impl.
impl<T: CounterBase> CounterServer for T {}
```

With the optional blanket impl, implementations cannot override
`CounterServer`'s behavior. This is similar to how
[`tokio::io::AsyncWriteExt`] has a blanket impl for
[`AsyncWrite`](tokio::io::AsyncWrite).

File: dropshot/examples/server-trait.rs [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+//! A basic example demonstrating use of the `dropshot::server` attribute macro
+//! to define a server.
+//!
+//! There are two parts: the interface and the implementation. The interface
+//! defines the endpoints and the types used by the server. The implementation
+//! provides the actual behavior of the server.
+//!
+//! In production code, the interface and implementation would likely be in
+//! separate crates. This example puts them in separate modules.
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply dropshot/examples/server-trait.rs:11]: Added an expanded note to `lib.rs`, and added a note about looking at that in this file.

File: dropshot/src/api_description.rs [hunk 9]
```diff
@@ -27,8 +29,16 @@ use serde::Serialize;
 use std::collections::BTreeMap;
 use std::collections::HashMap;
 use std::collections::HashSet;
+use std::fmt;
 use std::sync::Arc;
 
+/// A type used for stub implementations.
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply dropshot/src/api_description.rs:35]: Yeah, this is right. Thanks

File: dropshot/src/handler.rs [hunk 10]
```diff
@@ -494,6 +495,39 @@ where
     }
 }
 
+/// An unimplemented [`RouteHandler`] that panics when invoked.
+///
+/// The main use for this is when users would like to generate an OpenAPI spec
+/// for an endpoint that is not yet implemented.
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply dropshot/src/handler.rs:501]: Thanks!

File: dropshot/src/lib.rs [hunk 11]
```diff
@@ -112,9 +108,13 @@
...
-//! Typically, you define an endpoint with a handler function by using the
-//! [`endpoint`] macro. Here's an example of a single endpoint that lists
-//! a hardcoded project:
+//! There are two ways to define a set of endpoints:
+//!
+//! ### As a free function
+//!
+//! The simplest way to get started is by defining an endpoint as a top-level
+//! function, annotated with the [`endpoint`] macro. Here's an example of a
+//! single endpoint that lists a hardcoded project:
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply dropshot/src/lib.rs:117]: Done. Though "simplest" to me means "single", so I changed it to "an endpoint as a function".

File: dropshot/src/lib.rs [hunk 12]
```diff
@@ -178,6 +178,96 @@
 //! for each type of response (which can also include documentation).  This is
 //! largely known statically, though generated at runtime.
 //!
+//! ### As a trait server
+//!
+//! With Rust 1.75 and above, it is also possible to define an API using a
+//! *trait server*. A trait server is a trait that represents a collection
+//! of API endpoints. Each endpoint is defined as a static method on the trait,
+//! and the trait is annotated with `#[dropshot::server]`.
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply dropshot/src/lib.rs]: Thanks, moved the Rust version req to the end.

File: dropshot/src/lib.rs [hunk 13]
```diff
@@ -178,6 +178,96 @@
...
+//!
+//! One advantage of trait servers is that they can be defined in a separate
+//! crate from the implementation. This results in quicker builds for the API,
+//! and also make it easier to deal with changes to circular API dependencies.
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply dropshot/src/lib.rs]: Thanks, incorporated.

File: dropshot/src/lib.rs [hunk 14]
```diff
@@ -178,6 +178,96 @@
...
+//! use dropshot::ApiDescription;
+//! use dropshot::HttpError;
+//! use dropshot::HttpResponseOk;
+//! use dropshot::RequestContext;
+//! use http::Method;
+//! use schemars::JsonSchema;
+//! use serde::Serialize;
+//! use std::sync::Arc;
+//!
+//! /** Represents a project in our API */
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply dropshot/src/lib.rs]: Yeah I was wondering about that! Changed, and also changed the other uses for consistency.

File: dropshot/src/lib.rs [hunk 15]
```diff
@@ -337,6 +427,53 @@
 //! users commonize code using regular Rust functions and calling them.  See the
 //! design notes in the README for more on this.
 //!
+//! ### Generating OpenAPI descriptions
+//!
+//! For a given `ApiDescription`, you can also print out an OpenAPI spec
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply dropshot/src/lib.rs:463]: Changed, thanks.

File: dropshot/src/lib.rs [hunk 16]
```diff
@@ -178,6 +178,96 @@
...
+//!         project_server::api_description::<ServerImpl>().unwrap();
+//!
+//!     /* ... (use `api` to set up an `HttpServer` ) */
+//! }
+//! ```
+//!
+//! #### Limitations
+//!
+//! Currently, the `#[dropshot::server]` macro is only supported in module
+//! contexts, not function definitions. This is a Rust limitation -- see [Rust
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply dropshot/src/lib.rs]: That's right -- it's unfortunately not possible to have `#[dropshot::server]` inside function bodies. Changed "not function definitions" to "not within function bodies".

File: dropshot/src/lib.rs [hunk 17]
```diff
@@ -337,6 +427,53 @@
...
+//! # // defining fn main puts the doctest in a module context
+//! # fn main() {
+//! let description = project_server::stub_api_description().unwrap();
+//! let mut openapi = description.openapi("Project Server", "1.0.0");
+//! openapi.write(&mut std::io::stdout().lock()).unwrap();
+//! # }
+//! ```
+//!
+//! A stub description cannot be used for an actual server: all request handlers
+//! will panic.
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply dropshot/src/lib.rs]: Yes, it's possible to make a server out of it. Changed it to "must not".

File: dropshot_endpoint/src/params.rs [hunk 18]
```diff
@@ -138,15 +220,23 @@ impl<'ast> fmt::Debug for RqctxTy<'ast> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         match self {
             RqctxTy::Function(ty) => write!(f, "Function({})", quote! { #ty }),
+            RqctxTy::Trait { orig, transformed_unit } => {
+                write!(
+                    f,
+                    "Trait {{ orig: {}, transformed_unit: {} }}",
+                    quote! { #orig },
+                    quote! { #transformed_unit },
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply dropshot_endpoint/src/params.rs:228]: Oh I actually like `orig.to_token_stream`, not sure why I invoked the `quote!` macro here. Changed.

File: dropshot_endpoint/src/server.rs [hunk 19]
```diff
@@ -0,0 +1,1433 @@
...
+    let mut ret = String::new();
+    for (i, ch) in s.char_indices() {
+        if i > 0 && ch.is_uppercase() {
+            ret.push('_');
+        }
+        ret.push(ch.to_ascii_lowercase());
+    }
+
+    ret
+}
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply dropshot_endpoint/src/server.rs]: Ah neat -- I guess handling non-ASCII text is complex enough that I should just use heck here.

[the remark this answers is not in the document]
Comment (sunshowers) [reply dropshot_endpoint/src/server.rs]: Done.

File: dropshot_endpoint/src/server.rs [hunk 20]
```diff
@@ -0,0 +1,1433 @@
...
+        _cattr: &'ast syn::Attribute,
+    ) -> Result<Self, ServerItemErrorSummary> {
+        // TODO: implement channels
+        Ok(Self { f })
+    }
+
+    fn to_out_trait_item(&self) -> TraitItemFnForSignature {
+        // TODO
+        self.f.clone()
+    }
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply dropshot_endpoint/src/server.rs]: Yeah -- added a link to that PR.

File: dropshot_endpoint/src/syn_parsing.rs [hunk 21]
```diff
@@ -1,8 +1,15 @@
 // Copyright 2021 Oxide Computer Company
+//
+// Portions of this file are adapted from syn (https://github.com/dtolnay/syn),
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply dropshot_endpoint/src/syn_parsing.rs:3]: I've added "adapted from syn" comments to the relevant bits (which is most of the file).

File: dropshot_endpoint/src/server.rs [hunk 22]
```diff
@@ -0,0 +1,1433 @@
...
+                //
+                // 1. Put the functions below into the parent scope. This adds a
+                //    bunch of items to the scope rather than one, which seems
+                //    worse on balance.
+                // 2. Make these methods on a type rather than free functions in
+                //    a module. This approach works for functions but not other
+                //    items like macros we may wish to define in the future.
+                //
+                // In RFD 479, we determined that on balance, the current
+                // approach has the least practical downsides.
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply dropshot_endpoint/src/server.rs]: changed this to simply "fewest downsides".

File: dropshot/examples/server-trait-default.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (sunshowers) [reply dropshot/examples/server-trait-default.rs]: Done.

File: dropshot/examples/server-trait-default.rs
[the remark this answers is not in the document]
Comment (sunshowers) [reply dropshot/examples/server-trait-default.rs:1]: Changed to `api-trait-alternate.rs`.

**Pull request merged**